core/emulate.c

```diff
@@ -0,0 +1,899 @@
+        ctxt->dst.flags = 0;
+        opcode->decode_dst(ctxt, &ctxt->dst);
+    }
+    if (opcode->decode_src1) {
+        ctxt->src1.flags = 0;
+        opcode->decode_src1(ctxt, &ctxt->src1);
+    }
+    if (opcode->decode_src2) {
+        ctxt->src2.flags = 0;
+        opcode->decode_src2(ctxt, &ctxt->src2);
+    }
+}
+
+static void decode_op_none(em_context_t *ctxt,
+                           em_operand_t *op)
+{
+    op->type = OP_NONE;
+}
+
+static void decode_op_modrm_reg(em_context_t *ctxt,
+                                em_operand_t *op)
+{
+    op->type = OP_REG;
+    op->size = ctxt->operand_size;
+    op->reg.index = ctxt->modrm.reg | (ctxt->rex.r << 3);
+}
+
+static void decode_op_modrm_rm(em_context_t *ctxt,
+                               em_operand_t *op)
+{
+    uint32_t reg_base;
+    uint32_t reg_index;
+    uint8_t scale;
+
+    // Register operand
+    if (ctxt->modrm.mod == 3) {
+        op->type = OP_REG;
+        op->size = ctxt->operand_size;
+        op->reg.index = ctxt->modrm.rm | (ctxt->rex.b << 3);
+        return;
+    }
+
+    // Memory operand
+    op->type = OP_MEM;
+    op->size = ctxt->operand_size;
+    op->mem.ea = 0;
+    op->mem.seg = SEG_DS;
+    if (ctxt->override_segment) {
+        op->mem.seg = ctxt->override_segment;
+    }
+
+    if (ctxt->address_size == 2) {
+        /* Intel SDM Vol. 2A:
+         * Table 2-1. 16-Bit Addressing Forms with the ModR/M Byte */
+        switch (ctxt->modrm.rm) {
+        case 0:
+            op->mem.ea = BX + SI;
+            break;
+        case 1:
+            op->mem.ea = BX + DI;
+            break;
+        case 2:
+            op->mem.ea = BP + SI;
+            break;
+        case 3:
+            op->mem.ea = BP + DI;
+            break;
+        case 4:
+            op->mem.ea = SI;
+            break;
+        case 5:
+            op->mem.ea = DI;
+            break;
+        case 6:
+            if (ctxt->modrm.mod == 0) {
+                op->mem.ea = insn_fetch_u16(ctxt);
+            } else {
+                op->mem.ea = BP;
+            }
+            break;
+        case 7:
+            op->mem.ea = BX;
+            break;
+        }
+
+        // Displacement
+        if (ctxt->modrm.mod == 1) {
+            op->mem.ea += insn_fetch_s8(ctxt);
+        }
+        if (ctxt->modrm.mod == 2) {
+            op->mem.ea += insn_fetch_u16(ctxt);
+        }
+    }
+    if (ctxt->address_size == 4 || ctxt->address_size == 8) {
+        /* Intel SDM Vol. 2A:
+         * Table 2-2. 32-Bit Addressing Forms with the ModR/M Byte */
+        if (ctxt->modrm.rm == 4) {
+            /* Intel SDM Vol. 2A:
+             * Table 2-3. 32-Bit Addressing Forms with the SIB Byte */
+            ctxt->sib.value = insn_fetch_u8(ctxt);
+            reg_base  = ctxt->sib.base  | (ctxt->rex.b << 3);
+            reg_index = ctxt->sib.index | (ctxt->rex.x << 3);
+            scale = 1 << ctxt->sib.scale;
+            op->mem.ea += READ_GPR(reg_base, ctxt->address_size);
```

Comment:
There's a special case: when `mod == 0`, the base register is replaced by a 32-bit displacement. See the notes below Table 2-3 (Intel SDM Vol. 2A).

I've rewritten this part, for your reference:

```
            op->mem.ea = READ_GPR(reg_index, ctxt->address_size) * scale;
            if (ctxt->modrm.mod == 0) {
                op->mem.ea += insn_fetch_s32(ctxt);
            } else {
                uint32_t reg_base = ctxt->sib.base | (ctxt->rex.b << 3);
                op->mem.ea += READ_GPR(reg_base, ctxt->address_size);
            }
```

This seems to fix `debian_wheezy_amd64_standard.qcow2` boot!

Reply:
I got carried away and missed the `base == 5` check, haha. Thanks for fixing it!